### PR TITLE
Filter out unsatisfied nodes when calling  AddPod in PodTopologySpread

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes # https://github.com/kubernetes/kubernetes/issues/110726

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ahg-g** [comment]
/lgtm

I think the fix makes sense. 

/cc @alculquicondor who reviewed the node inclusion policy feature.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 1]
```diff
@@ -1501,13 +1503,67 @@ func TestPreFilterStateAddPod(t *testing.T) {
...
+					"zone": {{"zone2", 1}, {MatchNum: math.MaxInt32}},
+				},
+				TpPairToMatchNum: map[topologyPair]int{
+					{key: "zone", value: "zone2"}: 1,
+				},
+			},
+			enableNodeInclustionPolicy: false,
+		},
+		{
+			name: "add a pod with NotIn nodeAffinity when NodeInclustionPolicy enabled",
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1533]: Yes, nodeAffinity is honored, nodeTaint is ignored.

**Comment by kerthcet** [comment]
Updated and squashed.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 2]
```diff
@@ -1220,13 +1221,14 @@ func TestPreFilterStateAddPod(t *testing.T) {
...
-		nodes        []*v1.Node
-		want         *preFilterState
+		name                       string
+		preemptor                  *v1.Pod
+		addedPod                   *v1.Pod
+		existingPods               []*v1.Pod
+		nodeIdx                    int // denotes which node 'addedPod' belongs to
+		nodes                      []*v1.Node
+		want                       *preFilterState
+		enableNodeInclustionPolicy bool
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1231]: this makes me think that we need similar unit tests for TestPreFilterRemovePod (which could be done in a follow up).

Unless, we have a similar bug in RemovePod, then we should fix it in this PR.

/lgtm
/approve
/hold
to verify if RemovePod needs to be fixed.

Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1231]: Both `AddPod` and `RemovePod` will call `updateWithPod`, so it's fixed at the same time. I'll follow a new PR to cover the remove testcases.

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1231]: Thanks!

/hold cancel

**Comment by alculquicondor** [comment]
This is still marked as WIP, can you change that?

**Pull request merged**